File: dataflow_bench_in_cpp/predicate_op.cpp
// Copyright 2016 IDEXX
#include "dataflow_bench_in_cpp/predicate_op.h"

#include <cassert>

// Constructor
PredicateOp::PredicateOp(int predicate, OpInterface* next)
  : predicate_(predicate),
    next_(next) {
  assert(next);
}

void PredicateOp::invoke(int incoming) {
  if (incoming != predicate_) {
    next_->invoke(incoming);
  }
}

File: dataflow_bench_in_java/PredicateOp.java
public class PredicateOp implements OpInterface
{
    private int predicate;
    private OpInterface next;
    
    // Constructor
    public PredicateOp(int predicate, OpInterface next)
    {
        this.predicate = predicate;
        this.next = next;
    }
    
    @Override
    public void invoke(int incoming)
    {
        if (incoming != predicate) {
            next.invoke(incoming);
        }
    }
}
